builder_test.go

```diff
@@ -310,26 +310,187 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
+								Args:        []string{"some-other-arg"},
+								Direct:      true,
+								BuildpackID: "B",
+							},
 							{
 								Type:        "override-type",
 								Command:     "bpB-command",
 								Args:        []string{"bpB-arg"},
 								Direct:      false,
 								BuildpackID: "B",
 							},
+						}); s != "" {
+							t.Fatalf("Unexpected:\n%s\n", s)
+						}
+					})
+
+					it("should warn when overriding default process type, with non-default process type", func() {
+						bpA := testmock.NewMockBuildpack(mockCtrl)
+						buildpackStore.EXPECT().Lookup("A", "v1").Return(bpA, nil)
+						bpA.EXPECT().Build(gomock.Any(), config).Return(lifecycle.BuildResult{
+							Processes: []launch.Process{
+								{
+									Type:        "override-type",
+									Command:     "bpA-command",
+									Args:        []string{"bpA-arg"},
+									Direct:      true,
+									BuildpackID: "A",
+									Default:     true,
+								},
+								{
+									Type:        "second-override-type",
+									Command:     "some-command",
+									Args:        []string{"some-arg"},
+									Direct:      true,
+									BuildpackID: "A",
+									Default:     true,
+								},
+							},
+						}, nil)
+						bpB := testmock.NewMockBuildpack(mockCtrl)
+						buildpackStore.EXPECT().Lookup("B", "v2").Return(bpB, nil)
+						bpB.EXPECT().Build(gomock.Any(), config).Return(lifecycle.BuildResult{
+							Processes: []launch.Process{
+								{
+									Type:        "override-type",
+									Command:     "bpB-command",
+									Args:        []string{"bpB-arg"},
+									Direct:      false,
+									BuildpackID: "B",
+								},
+								{
+									Type:        "second-override-type",
+									Command:     "some-other-command",
+									Args:        []string{"some-other-arg"},
+									Direct:      true,
+									BuildpackID: "B",
+									Default:     true,
+								},
+							},
+						}, nil)
+
+						metadata, err := builder.Build()
+						if err != nil {
+							t.Fatalf("Unexpected error:\n%s\n", err)
+						}
+						if s := cmp.Diff(metadata.Processes, []launch.Process{
 							{
-								Type:        "some-other-type",
+								Type:        "override-type",
+								Command:     "bpB-command",
+								Args:        []string{"bpB-arg"},
+								Direct:      false,
+								BuildpackID: "B",
+								Default:     false,
+							},
+							{
+								Type:        "second-override-type",
 								Command:     "some-other-command",
 								Args:        []string{"some-other-arg"},
 								Direct:      true,
 								BuildpackID: "B",
+								Default:     true,
+							},
+						}); s != "" {
+							t.Fatalf("Unexpected:\n%s\n", s)
+						}
+
+						expected := "Warning: redefining the following default process types with processes not marked as default: [override-type]"
+						h.AssertStringContains(t, stdout.String(), expected)
+					})
+
+					it("should preserve ordering of processes", func() {
```

Comment:
I don't think this is necessary to be spec-compliant, though maybe it's good to have it?

Reply:
I see that you marked this conversation as resolved.
Do you still want to discuss it?

Reply:
I was confused as to why this was there before I saw `launch. FindLastDefaultProcessType`

It looks like the current implementation is always returning the processes sorted alphabetically: https://github.com/buildpacks/lifecycle/blob/fae1fe27c0acdfabe9538e3607e72af61c62ede9/builder.go#L186-L197

I am not sure, but I think the exporter _might_ be relying on this ordering when processing the layers. We should probably understand this better.

Reply:
I need to check the `sort.Strings` call but, as you wrote, the reason we introduced the test that is making sure that we're preserving the order, is since starting from `platform API 0.6` with `buildpack API 0.6`, we care about the order of the processes that sets the `default` field. For older APIs, we didn't need that because the user was able to set a default process only with passing the `-process-type` flag.

Reply:
It seems the current alphabetical ordering is required for reproducibility ([Slack convo](https://buildpacks.slack.com/archives/CSFE49VGR/p1612364432036100)). We should probably add a test for this, so that we don't accidentally change it in the future.

To communicate "last default process" to the exporter, we could add a new field in metadata.toml.